File: src/solution_java/medium/_150EvaluateReversePolishNotation.java
package solution_java.medium;

import java.util.ArrayDeque;

public class _150EvaluateReversePolishNotation {
    public class Solution {
        public int evalRPN(String[] tokens) {
            ArrayDeque<Integer> stack = new ArrayDeque<>();
            for (String ch : tokens) {
                if (isNumber(ch)) {
                    stack.push(Integer.parseInt(ch));
                } else {
                    int num1 = stack.pop();
                    int num2 = stack.pop();
                    switch (ch) {
                        case "+":
                            stack.push(num1 + num2);
                            break;
                        case "-":
                            stack.push(num2 - num1);
                            break;
                        case "*":
                            stack.push(num1 * num2);
                            break;
                        case "/":
                            stack.push(num2 / num1);
                            break;
                    }
                }
            }
            return stack.pop();
        }

        private boolean isNumber(String ch) {
            return !(ch.equals("+") || ch.equals("-") || ch.equals("*") || ch.equals("/"));
        }
    }
}

File: src/medium/_150EvaluateReversePolishNotation.kt
package medium

class _150EvaluateReversePolishNotation {
class Solution {
    fun evalRPN(tokens: Array<String>): Int {
        val stack = ArrayDeque<Int>()
        for (i in tokens.indices) {
            val item = tokens[i]
            if (isNumber(item)) {
                stack.add(item.toInt())
            } else {
                val num1 = stack.removeLast()
                val num2 = stack.removeLast()
                when (item) {
                    "+" -> stack.add(num2 + num1)
                    "-" -> stack.add(num2 - num1)
                    "*" -> stack.add(num2 * num1)
                    "/" -> stack.add(num2 / num1)
                    else -> {
                    }
                }
            }
        }
        return stack.removeLast()
    }

    private fun isNumber(str: String): Boolean {
        return !("+" == str || "-" == str || "*" == str || "/" == str)
    }
}
}